src/test/e2e/api/admin/segment.e2e.test.ts

```diff
@@ -0,0 +1,257 @@
+};
+
+const createSegment = (
+    postData: object,
+    expectStatusCode = 201,
+): Promise<unknown> => {
+    return app.request
+        .post(SEGMENTS_BASE_PATH)
+        .send(postData)
+        .expect(expectStatusCode);
+};
+
+const createFeatureToggle = (
+    postData: object,
+    expectStatusCode = 201,
+): Promise<unknown> => {
+    return app.request
+        .post(FEATURES_BASE_PATH)
+        .send(postData)
+        .expect(expectStatusCode);
+};
+
+const addSegmentToStrategy = (
+    segmentId: number,
+    strategyId: string,
+    expectStatusCode = 201,
+): Promise<unknown> => {
+    return app.request
+        .post(SEGMENTS_BASE_PATH + `/${segmentId}/strategies/${strategyId}`)
+        .set('Content-type', 'application/json')
+        .expect(expectStatusCode);
+};
+
+const mockFeatureToggle = () => {
+    return {
+        name: uuidv4(),
+        strategies: [{ name: uuidv4(), constraints: [], parameters: {} }],
+    };
+};
+
+const ids = <T>(items: { id?: T }[]): T[] => {
+    return items.map((item) => item.id);
+};
+
+beforeAll(async () => {
+    db = await dbInit('segments', getLogger);
+    app = await setupApp(db.stores);
+});
+
+afterAll(async () => {
+    await app.destroy();
+    await db.destroy();
+});
+
+afterEach(async () => {
+    await db.stores.segmentStore.deleteAll();
+    await db.stores.featureToggleStore.deleteAll();
+});
+
+test('should validate segments', async () => {
+    await createSegment({ something: 'a' }, 400);
+    await createSegment({ name: 'a', something: 'b' }, 400);
+    await createSegment({ name: 'a', constraints: 'b' }, 400);
+    await createSegment({ constraints: [] }, 400);
+    await createSegment({ name: 'a', constraints: [{}] }, 400);
+    await createSegment({ name: 'a', constraints: [] });
+});
+
+test('should create segments', async () => {
+    await createSegment({ name: 'a', constraints: [] });
+    await createSegment({ name: 'c', constraints: [] });
+    await createSegment({ name: 'b', constraints: [] });
+
+    const segments = await fetchSegments();
+    expect(segments.map((s) => s.name)).toEqual(['a', 'b', 'c']);
+});
+
+test('should update segments', async () => {
+    await createSegment({ name: 'a', constraints: [] });
+    const [segmentA] = await fetchSegments();
+    expect(segmentA.id).toBeGreaterThan(0);
+    expect(segmentA.name).toEqual('a');
+    expect(segmentA.createdAt).toBeDefined();
+    expect(segmentA.constraints.length).toEqual(0);
+
+    await app.request
+        .put(SEGMENTS_BASE_PATH)
+        .send({ ...segmentA, name: 'b' })
+        .expect(204);
+
+    const [segmentB] = await fetchSegments();
+    expect(segmentB.id).toEqual(segmentA.id);
+    expect(segmentB.name).toEqual('b');
+    expect(segmentB.createdAt).toBeDefined();
+    expect(segmentB.constraints.length).toEqual(0);
+});
+
+test('should update segment constraints', async () => {
+    const constraintA = { contextName: 'a', operator: 'IN' };
```

Comment:
I might just be reflecting my lack of knowledge of how we want this to work but this test suggests that we can make constraints with no value or values (I think this is wrong!) I think that's being caused by existing code not validating that the a constraint has a value xor values. Just something to think about in the coming weeks - if we patch that it'll break these tests

Reply:
Good point. I'll add `values` to the test data at least.